library/common/http/client.cc

```diff
@@ -112,19 +162,62 @@ void Client::DirectStreamCallbacks::encodeTrailers(const ResponseTrailerMap& tra
   ASSERT(http_client_.getStream(direct_stream_.stream_handle_));
   closeStream(); // Trailers always indicate the end of the stream.
 
+  // If in async mode, don't send data unless prompted.
+  if (async_mode_ && bytes_to_send_ == 0) {
+    response_trailers_ = ResponseTrailerMapImpl::create();
+    HeaderMapImpl::copyFrom(*response_trailers_, trailers);
+    return;
+  }
+
+  sendTrailersToBridge(trailers);
+}
+
+void Client::DirectStreamCallbacks::sendTrailersToBridge(const ResponseTrailerMap& trailers) {
   ENVOY_LOG(debug, "[S{}] dispatching to platform response trailers for stream:\n{}",
             direct_stream_.stream_handle_, trailers);
   bridge_callbacks_.on_trailers(Utility::toBridgeHeaders(trailers), bridge_callbacks_.context);
   onComplete();
 }
 
+void Client::DirectStreamCallbacks::resumeData(int32_t bytes_to_send) {
+  ASSERT(async_mode_);
+  ASSERT(bytes_to_send > 0);
+
+  bytes_to_send_ = bytes_to_send;
+
+  ENVOY_LOG(debug, "[S{}] received resume data call for {} bytes", direct_stream_.stream_handle_,
+            bytes_to_send_);
+
+  // If there is bufferd data, send up to bytes_to_send bytes.
+  // Make sure to send end stream with data only if
+  // 1) it has been received from the peer and
+  // 2) there are no trailers
+  if ((response_data_.get() && response_data_->length() != 0) ||
+      (end_stream_read_ && !end_stream_communicated_)) {
+    sendDataToBridge(*response_data_, end_stream_read_ && !response_trailers_.get());
+  }
+
+  // If all buffered data has been sent, send and free up trailers.
+  if ((!response_data_.get() || response_data_->length() == 0) && response_trailers_.get()) {
```

Comment:
nit: maybe add a function `hasBufferedResponseData` or something to make these if blocks easier to read?